File: Chapter3_VB/CustomerDAL.vb
Imports System.Data.SqlClient

Public Class CustomerDAL
    Private strConn As String
    Private conn As SqlConnection
    Private cmd As SqlCommand
    Private dr As SqlDataReader
    Private Customers As New List(Of Customer)

    Public Sub New()
        strConn = "Server=ACTUAL;Database=BobsShoes;Trusted_Connection=True;"
    End Sub

    Public Function GetCustomers() As List(Of Customer)
        Try
            Dim strSql = "SELECT * FROM Orders.Customers order by CustName"

            conn = New SqlConnection(strConn)
            cmd = New SqlCommand(strSql, conn)
            conn.Open()
            dr = cmd.ExecuteReader()
            If dr.HasRows Then
                While dr.Read
                    Dim cust As New Customer
                    cust.CustID = CInt(dr("CustID"))
                    cust.CustName = dr("CustName").ToString()
                    cust.CustStreet = dr("CustStreet").ToString()
                    cust.CustCity = dr("CustCity").ToString()
                    cust.CustStateProv = dr("CustStateProv").ToString()
                    cust.CustCountry = dr("CustCountry").ToString()
                    cust.CustPostalCode = dr("CustPostalCode").ToString()
                    cust.SalutationID = CInt(dr("SalutationID"))
                    Customers.Add(cust)
                End While
            End If
            dr.Close()

            Return Customers
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            conn.Close()
        End Try
    End Function

    Public Function GetCustomerWithDS() As DataSet
        Try
            Dim strSql = "SELECT * FROM Orders.Customers order by CustName"

            conn = New SqlConnection(strConn)
            cmd = New SqlCommand(strSql, conn)

            Dim ds As New DataSet
            Dim da As New SqlDataAdapter
            da.SelectCommand = cmd
            conn.Open()

            da.Fill(ds, "Customers")

            Return ds
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            conn.Close()
        End Try
    End Function
End Class

File: Chapter4/CustomerDAL.cs
using System.Data.SqlClient;

namespace SampleCSharp
{
    public class CustomerDAL
    {
        const string strConn = "Server=ACTUAL;Database=BobsShoes;Trusted_Connection=True;";

        public IEnumerable<Customer> GetAll()
        {
            List<Customer> customers = new List<Customer>();
            using (SqlConnection conn = new SqlConnection(strConn))
            {
                string strSql = @"SELECT * FROM Orders.Customers order by CustName";
                SqlCommand cmd = new SqlCommand(strSql, conn);
                conn.Open();
                SqlDataReader dr = cmd.ExecuteReader();

                if (dr.HasRows)
                {
                    while (dr.Read())
                    {
                        customers.Add(new Customer
                        {
                            CustID = Convert.ToInt32(dr["CustID"]),
                            CustName = dr["CustName"].ToString(),
                            CustStreet = dr["CustStreet"].ToString(),
                            CustCity = dr["CustCity"].ToString(),
                            CustStateProv = dr["CustStateProv"].ToString(),
                            CustCountry = dr["CustCountry"].ToString(),
                            CustPostalCode = dr["CustPostalCode"].ToString(),
                            SalutationID = Convert.ToInt32(dr["SalutationID"])
                        });
                    }
                }

                dr.Close();
                cmd.Dispose();
                conn.Close();

                return customers;
            }
        }
    }
}
